proposals/0000-dps-compact.md

```diff
@@ -0,0 +1,393 @@
+
+```haskell
+kind1TyVar, kind2TyVar :: TyVar
+(kind1TyVar : kind2TyVar : _)
+  = mkTemplateKindVars (repeat liftedTypeKind)
+
+kind1TyVarInf, kind2TyVarInf :: TyVarBinder
+kind1TyVarInf = mkTyVarBinder Inferred kind1TyVar
+kind2TyVarInf = mkTyVarBinder Inferred kind2TyVar
+
+kind1Ty, kind2Ty :: Type
+kind1Ty = mkTyVarTy kind1TyVar
+kind2Ty = mkTyVarTy kind2TyVar
+
+kindPolyAlphaTyVar, kindPolyBetaTyVar :: TyVar
+[kindPolyAlphaTyVar, kindPolyBetaTyVar] =
+  mkTemplateTyVars
+    [ kind1Ty
+    , kind2Ty]
+
+kindPolyAlphaTyVarSpec, kindPolyBetaTyVarSpec :: TyVarBinder
+kindPolyAlphaTyVarSpec = mkTyVarBinder Specified kindPolyAlphaTyVar
+kindPolyBetaTyVarSpec  = mkTyVarBinder Specified kindPolyBetaTyVar
+
+kindPolyAlphaTy, kindPolyBetaTy :: Type
+kindPolyAlphaTy = mkTyVarTy kindPolyAlphaTyVar
+kindPolyBetaTy  = mkTyVarTy kindPolyBetaTyVar
+```
+
+### Obtain data constructor name as a `Symbol`: `LiftedCtorToSymbol` built-in type family
+
+```haskell
+type family LiftedCtorToSymbol (liftedCtor :: k) :: Symbol
+```
+
+When making an API around `compactAddHollow#`, we need to know the number and type of the fields that the constructor has. That information can be easily retrieved with `GHC.Generics`. However, the type-level structure returned by `GHC.Generics` distinguishes constructors of a given type by their name, as a `Symbol`.
+
+I introduce a built-in type family `LiftedCtorToSymbol` whose in charge of translating a data constructor lifted as a type into the corresponding symbol. That way a library can leverage `GHC.Generics` to allow direct and safe structure building in compact regions for most user-defined data types.
+
+For example:
+
+```haskell
+LiftedCtorToSymbol 'Left ~ "Left"
+```
+
+More generally, forall data constructor `〈d〉`,
+
+```haskell
+LiftedCtorToSpecCtor '〈d〉 ~ "〈d〉"
+```
+
+## Proposed Library Change Specification
+
+- `compactAddHollow#`, `reifyInfoTablePtr#`, and `InfoTablePtrOf#` are exposed in `GHC.Prim`
+- `LiftedCtorToSymbol` is exposed in `GHC.TypeLits`
+
+## Examples
+
+### Primitive usage
+
+#### Allocate a hollow `Left` constructor in region `compactRegion`
+
+```haskell
+let leftInfoPtr :: InfoTablePtrOf# 'Left = reifyInfoTablePtr# (# #)
+ in compactAddHollow# compactRegion (unsafeCoerceAddr leftInfoPtr) state
+```
+
+Yet unspecified fields can later be assigned with existing primop `writeAddrOffAddr#` (we know the offsets of the fields thanks to the return type of `compactAddHollow#`).
+
+#### Allocate hollow `stg_IND` object in region `compactRegion`
+
+```haskell
+let indInfoPtr :: InfoTablePtrOf# "IND" = reifyInfoTablePtr# (# #)
+ in compactAddHollow# compactRegion (unsafeCoerceAddr indInfoPtr) state
+```
+
+#### Get info about fields of a data constructor using `GHC.Generics` and `LiftedCtorToSymbol`
```

Comment:
This is a hefty example. It takes a while to even find where the stuff from the proposal shows up. Can we find a shorter example? (maybe the same example but skipping some cases?)

Reply:
I cut down the example a bit, and added comments. I think it reads better now, let me know :)